// purify-ts/esm/Function.js
/** The identity function, returns the value it was given */
export const identity = (x) => x;
/** Returns a function that always returns the same value. Also known as `const` in other languages */
export const always = (x) => () => x;
export const Order = {
    LT: 'LT',
    EQ: 'EQ',
    GT: 'GT'
};
/** Compares two values using the default "<" and ">" operators */
export const compare = (x, y) => {
    if (x > y) {
        return Order.GT;
    }
    else if (x < y) {
        return Order.LT;
    }
    else {
        return Order.EQ;
    }
};
/** Maps the Order enum to the values expected by the standard ECMAScript library when doing comparison (Array.prototype.sort, for example) */
export const orderToNumber = (order) => {
    switch (order) {
        case Order.LT:
            return -1;
        case Order.EQ:
            return 0;
        case Order.GT:
            return 1;
    }
};
/** Takes a function that receives multiple arguments and returns a "curried" version of that function that can take any number of those arguments and if they are less than needed a new function that takes the rest of them will be returned */
export const curry = (fn) => function currify(...args) {
    return args.length >= fn.length
        ? fn.apply(undefined, args)
        : currify.bind(undefined, ...args);
};
